### PR TITLE
add database pkg from moov-io/identity

### PR DESCRIPTION
**Pull request by vxio**
Related issue: https://github.com/moov-io/identity/issues/60

### CODE REVIEW

**Comment by adamdecaf** [comment]
Did this fail because there were no migrations to run?

**Comment by vxio** [comment]
I'm not sure why that test failed; ran fine on my machine.

**Comment by adamdecaf** [comment]
Restarted that test. Let's see what happens.

**Comment by vxio** [comment]
@adamdecaf you were right, builds were failing because of no migrations folder in `base` repo; added a quick fix.

**Comment by vxio** [comment]
Builds still failing with:
`dial tcp [::1]:4306: connect: connection refused--- FAIL: Test_NewAndMigration_MySql (0.00s)`

I think Docker Build isn't setup for this repo yet.

**Comment by adamdecaf** [comment]
We probably need to setup docker image launching in this test. Identity diverted from Customers/PayGate in requiring `docker-compose up` rather than using `dockertest`. 

https://github.com/moov-io/customers/blob/master/internal/database/mysql.go#L222

File: database/mysql.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, err
+	}
+	db.SetMaxOpenConns(maxActiveMySQLConnections)
+
+	// Check out DB is up and working
+	if err := db.Ping(); err != nil {
+		return nil, err
+	}
+
+	// Migrate our database
```
Comment (alovak) [inline database/mysql.go:70]: remove this line?

File: database/migrator.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	log.Info().Log("Running Migrations")
+	_ = pkger.Include(config.migrationsDir)
+
+	driver, err := GetDriver(db, config)
+	if err != nil {
+		return err
+	}
+
+	m, err := migrate.NewWithDatabaseInstance(
+		"pkger:///migrations/",
```
Comment (alovak) [inline database/migrator.go:32]: This one should use `config.migrationsDir` (most probably `path.Base(config.migrationsDir)`)

From migrate package:

```go
pkger.Include("/module/path/to/migrations")
m, err := migrate.New("pkger:///module/path/to/migrations", ...)
```

File: database/database_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	config := DatabaseConfig{
+		DatabaseName: "identity",
+		MySql: &MySqlConfig{
+			Address:  "tcp(localhost:4306)",
+			User:     "identity",
+			Password: "identity",
+		},
+	}
+
+	_, close, err := NewAndMigrate(config, nil, nil)
```
Comment (alovak) [inline database/database_test.go:33]: I'm curious how does it work if we don't have mysql instance running and we do not expect error here?

File: database/mysql.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+func (r *TestMySQLDB) Close() error {
+	r.shutdown()
+
+	// Verify all connections are closed before closing DB
+	if conns := r.DB.Stats().OpenConnections; conns != 0 {
+		panic(fmt.Sprintf("found %d open MySQL connections", conns))
+	}
+
+	r.container.Close()
```
Comment (alovak) [inline database/mysql.go:123]: I would still close container even if we have opened connections.

File: database/sqlite.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+	db, err := sql.Open("sqlite3", s.path)
+	if err != nil {
+		return nil, err
+	}
+	if err := db.Ping(); err != nil {
+		return db, err
+	}
+
+	// Migrate our database
```
Comment (alovak) [inline database/sqlite.go:59]: ✂ comment

File: database/database_test.go (hunk 3 above)
Comment (alovak) [reply database/database_test.go:33]: I missed your comment on this.

**Review by adamdecaf — APPROVED** [review]

**Comment by adamdecaf** [comment]
Looks good, but we're still failing because MySQL isn't running in tests. 

https://github.com/moov-io/base/pull/83#issuecomment-698612913

**Pull request merged**